### PR TITLE
Feature: Add VE.Direct "RELAY" to live view, MQTT, and HASS

### PR DESCRIPTION
**Pull request by SW-Niko**
- process the "RELAY" information from the charger
- add to live view
- add to MQTT and HASS
Note: Only processed if available. Not all charger contain a fault relais

### CODE REVIEW

**Comment by SW-Niko** [comment]
#1037 
Note: I did not implement "FWE" because according the Victron information "FWE" is only for "Phoenix Charger" and "Smart BuckBoost" and not for MPPT.

@schlimmchen : I do not have the right charger. Someone else must do the testing.

File: src/MqttHandleVedirectHass.cpp [hunk 1]
```diff
@@ -88,6 +88,9 @@ void MqttHandleVedirectHassClass::publishConfig()
         publishSensor("Panel maximum power yesterday", NULL, "H23", "power", "measurement", "W", *optMpptData);
 
         // optional info, provided only if the charge controller delivers the information
+        if (optMpptData->relayState_RELAY.first != 0) {
+            publishBinarySensor("MPPT error relay state", "mdi:export", "RELAY", "ON", "OFF", *optMpptData);
```
Comment (schlimmchen) [inline src/MqttHandleVedirectHass.cpp:92]: Please use `mdi:electric-switch` as an icon: 
image

**Review by schlimmchen — CHANGES_REQUESTED** [review]
Thanks for this. The icon for HASS auto discovery is not a good choice, but otherwise this looks good. I can (mostly) test this with my father's charge controller.

**Comment by schlimmchen** [comment]
> Note: I did not implement "FWE" because according the Victron information "FWE" is only for "Phoenix Charger" and "Smart BuckBoost" and not for MPPT.

Can I motivate you to process and ignore "FWE" values and log a verbose message "FWE ignored"? I agree that we should ignore this, as it should not be there at all and if so, I expect it to be redundant to the other firmware info field. The benefit of explicitly ignoring the value is that we made the decision to ignore it and we make that clear by emitting a respective message, instead of suggesting that we don't know what FWE is. And we could then close #1037 as a whole :wink:

**Comment by SW-Niko** [comment]
No problem. Can I do this weekend 😄

File: src/MqttHandleVedirectHass.cpp (hunk 1 above)
Comment (SW-Niko) [reply src/MqttHandleVedirectHass.cpp]: Done

**Comment by SW-Niko** [comment]
Now we can get the firmware version from FW or FWE. 
All high level functions do not see any difference if version was transmitted by FE or FWE.
I think we can close #1037

File: lib/VeDirectFrameHandler/VeDirectFrameHandler.cpp [hunk 2]
```diff
@@ -253,6 +253,15 @@ void VeDirectFrameHandler<T>::processTextData(std::string const& name, std::stri
 		return;
 	}
 
+    // Some devices use "FWE" instead of "FW" for the firmware version.
+    // We just remove the last 2 digit (final or beta indication) and copy the FW number.
+	if (name == "FWE") {
+        if ((value.length() - 2) <= sizeof(_tmpFrame.firmwareVer_FW)) {
```
Comment (schlimmchen) [inline lib/VeDirectFrameHandler/VeDirectFrameHandler.cpp:259]: So if the device does not report FW, we still ignore FWE... I don't understand this if clause... We expect that the Victron device either sends only FW xor FWE, or that the two are equal. So we should process FWE unconditionally.

Have you checked what happens if the FWE string has a leading zero, which is optional:
image

**Review by schlimmchen — CHANGES_REQUESTED** [review]

File: lib/VeDirectFrameHandler/VeDirectFrameHandler.cpp (hunk 2 above)
Comment (schlimmchen) [reply lib/VeDirectFrameHandler/VeDirectFrameHandler.cpp:259]: FYI: the line below, the use of strncpy, contains a possible write out of bounds. you cannot assume that the value's string minus two chars fits the target variable...

I will take care of that as well.

**Comment by schlimmchen** [comment]
Tested successfully using https://github.com/schlimmchen/vedirect-mppt-emulator

**Comment by SW-Niko** [comment]
Hello @schlimmchen ,
thanks for fixing my work. As always ... your work ist much more better. 👍

**Pull request merged**